File: Python/82_Remove_Duplicates_from_Sorted_List_II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if head == None:
            return None
        return self.deleteDuplicates1(head)
    
    def deleteDuplicates1(self, head, h=None):
        if head.next == None:
            if h == None or head.val != h.val:
                return head
            return None
        if head.val != head.next.val and (h == None or head.val != h.val):
            head.next = self.deleteDuplicates1(head.next)
            return head
        else:
            return self.deleteDuplicates1(head.next, head)

File: C++/82_Remove_Duplicates_from_Sorted_List_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(head == NULL) return NULL;
        return deleteDuplicates1(head);
    }
    ListNode* deleteDuplicates1(ListNode* head, ListNode* h = NULL){
        if(head->next == NULL){
            if(h == NULL || head->val != h->val){
                return head;
            }
            return NULL;
        }
        if(head->val != head->next->val && (h == NULL || head->val != h->val)){
            head->next = deleteDuplicates1(head->next);
            return head;
        } else {
            return deleteDuplicates1(head->next, head);
        }
        return NULL;
    }
};